src/view/com/composer/Composer.tsx

```diff
@@ -625,10 +625,108 @@ export function useComposerCancelRef() {
   return useRef<CancelRef>(null)
 }
 
+function useAnimatedBorders() {
+  const t = useTheme()
+  const hasScrolledTop = useSharedValue(0)
+  const hasScrolledBottom = useSharedValue(0)
+  const contentOffset = useSharedValue(0)
+  const scrollViewHeight = useSharedValue(Infinity)
+  const contentHeight = useSharedValue(0)
+
+  /**
+   * Make sure to run this on the UI thread!
+   */
+  const showHideBottomBorder = useCallback(
+    ({
+      newContentHeight,
+      newContentOffset,
+      newScrollViewHeight,
+    }: {
+      newContentHeight?: number
+      newContentOffset?: number
+      newScrollViewHeight?: number
+    }) => {
+      'worklet'
+
+      if (typeof newContentHeight === 'number')
+        contentHeight.value = newContentHeight
+      if (typeof newContentOffset === 'number')
+        contentOffset.value = newContentOffset
+      if (typeof newScrollViewHeight === 'number')
+        scrollViewHeight.value = newScrollViewHeight
+
+      hasScrolledBottom.value = withTiming(
+        contentHeight.value - contentOffset.value >= scrollViewHeight.value
+          ? 1
+          : 0,
+      )
+    },
+    [contentHeight, contentOffset, scrollViewHeight, hasScrolledBottom],
+  )
+
+  const scrollHandler = useAnimatedScrollHandler({
+    onScroll: event => {
+      hasScrolledTop.value = withTiming(event.contentOffset.y > 0 ? 1 : 0)
+
+      // already on UI thread
+      showHideBottomBorder({
+        newContentOffset: event.contentOffset.y,
+        newContentHeight: event.contentSize.height,
+        newScrollViewHeight: event.layoutMeasurement.height,
+      })
+    },
+  })
+
+  const onScrollViewContentSizeChange = useCallback(
+    (_width: number, height: number) => {
+      runOnUI(showHideBottomBorder)({
+        newContentHeight: height,
+      })
+    },
+    [showHideBottomBorder],
+  )
+
+  const onScrollViewLayout = useCallback(
+    (evt: LayoutChangeEvent) => {
+      runOnUI(showHideBottomBorder)({
+        newScrollViewHeight: evt.nativeEvent.layout.height,
+      })
+    },
+    [showHideBottomBorder],
+  )
```

Comment:
In these cases, the preferred method seems to be to just update the `SharedValue` from this callback, and make `hasScrolledBottom` a derived value with `useDerivedValue`. Makes it a little easier and lets us avoid the `showHideBottomBorder` function all together.

Reply:
fwiw i remember bumping into issues with useDerivedValue in the past and having to rewrite those but i don’t remember why. so just worth checking that it works well if we switch it

Reply:
ooo, yea do check them. it shouldn't necessarily matter since either from the JS thread is async so if this is more reliable it seems fine, unsure if the suggestion in their docs is just for code cleanliness or if there's some other reason as well.

Reply:
I didn't see a way to derive a value from multiple inputs?

Reply:
_Should_ just be able to use it like so, and the returned value will update once one or multiple of the shared values changes

```tsx
const hasScrolledBottom = useDerivedValue(() => {
    return withTiming(
      contentHeight.value - contentOffset.value > scrollViewHeight.value ? 1 : 0,
    )
  })
```